drivers/adc/adc_nrfx_saadc.c

```diff
@@ -197,281 +204,342 @@ static int saadc_pm_hook(const struct device *dev, enum pm_device_action action)
+		LOG_ERR("Cannot set reference: %d", err);
+		return -EINVAL;
+	}
+
+	err = acq_time_set(ch_cfg, channel_cfg->acquisition_time);
+	if (err != 0) {
+		LOG_ERR("Cannot set acquisition time: %d", err);
 		return -EINVAL;
 	}
 
 	/* Store channel mode to allow correcting negative readings in single-ended mode
 	 * after ADC sequence ends.
 	 */
 	if (channel_cfg->differential) {
-		config.mode = NRF_SAADC_MODE_DIFFERENTIAL;
+		ch_cfg->mode = NRF_SAADC_MODE_DIFFERENTIAL;
 		m_data.single_ended_channels &= ~BIT(channel_cfg->channel_id);
 	} else {
-		config.mode = NRF_SAADC_MODE_SINGLE_ENDED;
+		ch_cfg->mode = NRF_SAADC_MODE_SINGLE_ENDED;
 		m_data.single_ended_channels |= BIT(channel_cfg->channel_id);
 	}
 
-#if (NRF_SAADC_HAS_AIN_AS_PIN)
-	if ((channel_cfg->input_positive >= ARRAY_SIZE(saadc_psels)) ||
-	    (channel_cfg->input_positive < NRF_SAADC_AIN0)) {
+	nrfx_err_t ret = nrfx_saadc_channel_config(&cfg);
+
+	if (ret != NRFX_SUCCESS) {
+		LOG_ERR("Cannot configure channel %d: %d", channel_cfg->channel_id, ret);
 		return -EINVAL;
 	}
 
-	if (config.mode == NRF_SAADC_MODE_DIFFERENTIAL) {
-		if (input_negative > NRF_SAADC_AIN7 ||
-		    input_negative < NRF_SAADC_AIN0) {
-			return -EINVAL;
-		}
-
-		input_negative = saadc_psels[input_negative];
-	} else {
-		input_negative = NRF_SAADC_INPUT_DISABLED;
-	}
-#endif
-	/* Store the positive input selection in a dedicated array,
-	 * to get it later when the channel is selected for a sampling
-	 * and to mark the channel as configured (ready to be selected).
-	 */
-	m_data.positive_inputs[channel_id] = channel_cfg->input_positive;
-
-	nrf_saadc_channel_init(NRF_SAADC, channel_id, &config);
-	/* Keep the channel disabled in hardware (set positive input to
-	 * NRF_SAADC_INPUT_DISABLED) until it is selected to be included
-	 * in a sampling sequence.
-	 */
-	nrf_saadc_channel_input_set(NRF_SAADC,
-				    channel_id,
-				    NRF_SAADC_INPUT_DISABLED,
-				    input_negative);
-
 	return 0;
 }
 
 static void adc_context_start_sampling(struct adc_context *ctx)
 {
 #if defined(CONFIG_PM_DEVICE_RUNTIME)
 	pm_device_runtime_get(DEVICE_DT_INST_GET(0));
-#else
-	nrf_saadc_enable(NRF_SAADC);
 #endif
 
 	if (ctx->sequence.calibrate) {
-		nrf_saadc_task_trigger(NRF_SAADC,
-				       NRF_SAADC_TASK_CALIBRATEOFFSET);
+		nrfx_saadc_offset_calibrate(event_handler);
 	} else {
-		nrf_saadc_task_trigger(NRF_SAADC, NRF_SAADC_TASK_START);
-		nrf_saadc_task_trigger(NRF_SAADC, NRF_SAADC_TASK_SAMPLE);
+		nrfx_err_t ret = nrfx_saadc_mode_trigger();
+
+		if (ret != NRFX_SUCCESS) {
+			LOG_ERR("Cannot start sampling: %d", ret);
+		}
 	}
 }
 
-static void adc_context_update_buffer_pointer(struct adc_context *ctx,
-					      bool repeat)
+static void adc_context_update_buffer_pointer(struct adc_context *ctx, bool repeat)
 {
-	ARG_UNUSED(ctx);
-
 	if (!repeat) {
 #if defined(ADC_BUFFER_IN_RAM)
 		m_data.user_buffer = (uint8_t *)m_data.user_buffer +
 			samples_to_bytes(&ctx->sequence, nrfy_saadc_amount_get(NRF_SAADC));
```

Comment:
can't we get rid of this `nrfy_saadc_amount_get` given that we have access to `m_data.event` ?